### PR TITLE
wip adding a gcs artifact type

### PR DESCRIPTION
**Pull request by decarboxy**
This PR adds support for using GCS as an artifact repository natively (rather than using S3 interoperability keys). 

In order for it to work,  the environment variable `GOOGLE_CREDENTIALS_SECRET` must be set in the workflow controller pod with the name of the k8s secret containing default application credentials.  This has been tested and is at this point ready for review

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -113,6 +113,15 @@ func (in *ArtifactLocation) DeepCopyInto(out *ArtifactLocation) {
 			**out = **in
 		}
 	}
+	if in.GCS != nil {
+		in, out := &in.GCS, &out.GCS
+		if *in == nil {
+			*out = nil
+		} else {
+			*out = new(GCSArtifact)
+			**out = **in
```
Comment (gaigepr) [inline pkg/apis/workflow/v1alpha1/zz_generated.deepcopy.go:122]: And here we see C's influence on go 🤔

Comment (decarboxy) [reply pkg/apis/workflow/v1alpha1/zz_generated.deepcopy.go:122]: its generated code, not my fault 😬

File: api/openapi-spec/swagger.json [hunk 2]
```diff
@@ -179,6 +187,32 @@
         }
       }
     },
+    "io.argoproj.workflow.v1alpha1.GCSArtifact": {
+      "description": "GCSBucket is the location of a GCS artifact",
```
Comment (guerard) [inline api/openapi-spec/swagger.json]: s/GCSBucket/GCSArtifact

File: pkg/apis/workflow/v1alpha1/openapi_generated.go [hunk 3]
```diff
@@ -298,6 +310,46 @@ func GetOpenAPIDefinitions(ref common.ReferenceCallback) map[string]common.OpenA
 			Dependencies: []string{
 				"github.com/argoproj/argo/pkg/apis/workflow/v1alpha1.DAGTask"},
 		},
+		"github.com/argoproj/argo/pkg/apis/workflow/v1alpha1.GCSArtifact": {
+			Schema: spec.Schema{
+				SchemaProps: spec.SchemaProps{
+					Description: "GCSBucket is the location of a GCS artifact",
```
Comment (guerard) [inline pkg/apis/workflow/v1alpha1/openapi_generated.go]: s/GCSBucket/GCSArtifact

File: workflow/artifacts/gcs/gcs.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+}
+
+func (gcsDriver *GCSArtifactDriver) saveToFile(inputArtifact *wfv1.Artifact, filePath string) error {
+
+	log.Infof("Loading from GCS (gs://%s/%s) to %s",
+		inputArtifact.GCS.Bucket, inputArtifact.GCS.Key, filePath)
+
+	stat, err := os.Stat(filePath)
+	if err == nil {
```
Comment (guerard) [inline workflow/artifacts/gcs/gcs.go]: nit: for readability push up the first `err != nil` before this line, and then you don't need to test `err == nil` explicitly.

or use '&&' so it's a single conditional.

just a little awkward as-is w/ the nested block, and then `err != nil` below

File: workflow/artifacts/gcs/gcs.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		return err
+	}
+
+	gcsClient, err := gcsDriver.newGcsClient()
+	if err != nil {
+		return err
+	}
+
+	bucket := gcsClient.Bucket(inputArtifact.GCS.Bucket)
```
Comment (guerard) [inline workflow/artifacts/gcs/gcs.go]: I'm assuming we don't have to do any empty-string checks due to how the serialization works (else you should add them).

File: workflow/artifacts/gcs/gcs.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		}
+	}
+
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+	}
+
+	outputFile, err := os.OpenFile(filePath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
```
Comment (guerard) [inline workflow/artifacts/gcs/gcs.go]: blocker: Don't you want to use `O_TRUNC` and truncate anything in the file if it exists? I assume you want the GCS artifact to overwrite any existing contents. Can remove `O_APPEND` in that case.

File: workflow/artifacts/gcs/gcs.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+func (gcsDriver *GCSArtifactDriver) saveToGCS(outputArtifact *wfv1.Artifact, filePath string) error {
+
+	log.Infof("Saving to GCS (gs://%s/%s)",
+		outputArtifact.GCS.Bucket, outputArtifact.GCS.Key)
+
+	gcsClient, err := gcsDriver.newGcsClient()
+	if err != nil {
+		return err
+	}
+
```
Comment (guerard) [inline workflow/artifacts/gcs/gcs.go:82]: q: any reason you don't have stat.isDir check in this function? Can we assume something here that we can't in saveToFile?

File: workflow/artifacts/gcs/gcs.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+func (gcsDriver *GCSArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
+
+	err := gcsDriver.saveToFile(inputArtifact, path)
```
Comment (guerard) [inline workflow/artifacts/gcs/gcs.go:118]: you can just return the value of the call directly (no need for local var, nil-check, etc.)

File: workflow/artifacts/gcs/gcs.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (gcsDriver *GCSArtifactDriver) Save(path string, outputArtifact *wfv1.Artifact) error {
+
+	err := gcsDriver.saveToGCS(outputArtifact, path)
```
Comment (guerard) [inline workflow/artifacts/gcs/gcs.go:124]: ditto

**Review by guerard — CHANGES_REQUESTED** [review]
Couple typos, and one file I/O blocker, but otherwise lgtm.

File: workflow/artifacts/gcs/gcs.go (hunk 6 above)
Comment (decarboxy) [reply workflow/artifacts/gcs/gcs.go]: yeah definitely

Comment (decarboxy) [reply workflow/artifacts/gcs/gcs.go]: not sure what i was thinking there, good catch

File: workflow/artifacts/gcs/gcs.go (hunk 7 above)
Comment (decarboxy) [reply workflow/artifacts/gcs/gcs.go:82]: thats an oversight, i'll add the isDir check

**Review by guerard — APPROVED** [review]
```
              |    |    |
             )_)  )_)  )_)
            )___))___))___)\
           )____)____)_____)\\
         _____|____|____|____\\\__
---------\                   /---------
  ^^^^^ ^^^^^^^^^^^^^^^^^^^^^
    ^^^^      ^^^^     ^^^    ^^
         ^^^^      ^^^
```
LGTM
